src/librbd/crypto/CryptoObjectDispatch.cc

```diff
@@ -0,0 +1,197 @@
+    const ::SnapContext &snapc, int op_flags,
+    const ZTracer::Trace &parent_trace, int* object_dispatch_flags,
+    uint64_t* journal_tid, io::DispatchResult* dispatch_result,
+    Context** on_finish, Context* on_dispatched) {
+  auto cct = m_image_ctx->cct;
+  ldout(cct, 20) << data_object_name(m_image_ctx, object_no) << " "
+                 << object_off << "~" << data.length() << dendl;
+
+  m_image_ctx->crypto->encrypt(
+          std::move(data),
+          get_image_offset(m_image_ctx, object_no, object_off));
+  *dispatch_result = io::DISPATCH_RESULT_CONTINUE;
+  on_dispatched->complete(0);
+  return true;
+}
+
+template <typename I>
+bool CryptoObjectDispatch<I>::write_same(
+    uint64_t object_no, uint64_t object_off, uint64_t object_len,
+    io::LightweightBufferExtents&& buffer_extents, ceph::bufferlist&& data,
+    const ::SnapContext &snapc, int op_flags,
+    const ZTracer::Trace &parent_trace, int* object_dispatch_flags,
+    uint64_t* journal_tid, io::DispatchResult* dispatch_result,
+    Context** on_finish, Context* on_dispatched) {
+  auto cct = m_image_ctx->cct;
+  ldout(cct, 20) << data_object_name(m_image_ctx, object_no) << " "
+                 << object_off << "~" << object_len << dendl;
+
+  // convert to regular write
+  io::LightweightObjectExtent extent(object_no, object_off, object_len, 0);
+  extent.buffer_extents = std::move(buffer_extents);
+
+  bufferlist ws_data;
+  io::util::assemble_write_same_extent(extent, data, &ws_data, true);
+
+  auto ctx = new LambdaContext(
+      [on_finish_ctx=on_dispatched](int r) {
+          on_finish_ctx->complete(r);
+      });
+
+  *dispatch_result = io::DISPATCH_RESULT_COMPLETE;
+  auto req = io::ObjectDispatchSpec::create_write(
+          m_image_ctx, io::OBJECT_DISPATCH_LAYER_NONE, object_no,
+          object_off, std::move(ws_data), snapc, op_flags, 0,
+          parent_trace, ctx);
+  req->send();
+  return true;
+}
+
+template <typename I>
+bool CryptoObjectDispatch<I>::compare_and_write(
+    uint64_t object_no, uint64_t object_off, ceph::bufferlist&& cmp_data,
+    ceph::bufferlist&& write_data, const ::SnapContext &snapc, int op_flags,
+    const ZTracer::Trace &parent_trace, uint64_t* mismatch_offset,
+    int* object_dispatch_flags, uint64_t* journal_tid,
+    io::DispatchResult* dispatch_result, Context** on_finish,
+    Context* on_dispatched) {
+  auto cct = m_image_ctx->cct;
+  ldout(cct, 20) << data_object_name(m_image_ctx, object_no) << " "
+                 << object_off << "~" << write_data.length()
+                 << dendl;
+
+  uint64_t image_offset = get_image_offset(m_image_ctx, object_no, object_off);
+  m_image_ctx->crypto->encrypt(std::move(cmp_data), image_offset);
+  m_image_ctx->crypto->encrypt(std::move(write_data), image_offset);
+  *dispatch_result = io::DISPATCH_RESULT_CONTINUE;
+  on_dispatched->complete(0);
+  return true;
+}
+
+template <typename I>
+uint64_t CryptoObjectDispatch<I>::get_image_offset(
```

Comment:
If this is attempting to perform full-image LUKS-like encryption, perhaps there really should be two different crypto plugins instead of trying to fit them both into the object-level encryption solution.

Reply:
CryptoInterface seems generic enough to me, as I assume most of not all realistic encryption modes will require the image offset as a parameter.

Reply:
But you're right about that from a cryptographic point of view, it should not matter if we replace the image offset by some 1-to-1 mapping like (object_no, object_off).